### PR TITLE
backports for ActivityLog and Reporting 1.11.x

### PR DESCRIPTION
**Pull request by mpalmi**
This PR encapsulates the backports for Census and ActivityLog changes to get the release branches in sync and should resolve [VAULT-14702](https://hashicorp.atlassian.net/browse/VAULT-14702) and [VAULT-14765](https://hashicorp.atlassian.net/browse/VAULT-14765) along with the backports for the new client count testing library.

It was easier to backport the changes en masse, since they're all intertwined. Teasing apart feature-specific changes would have been more difficult.

Commits were prepared using (with fish shell, $() is just ()):

```sh
git cherry-pick -v -x (git log origin/release/1.11.x..origin/main --oneline --reverse -- "*activity*" | awk '{ print $1 }')
```

## Diff between backport and main

```sh
## NOTE: This diff does not include changes to core. 
git diff origin/activitylog-backports-1.11.x..origin/main -- "*activity*" "*census*" ':!ui/*'
```

https://gist.github.com/mpalmi/9b83c19a45002a045b216d6d6446f357

[VAULT-14702]: https://hashicorp.atlassian.net/browse/VAULT-14702?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[VAULT-14765]: https://hashicorp.atlassian.net/browse/VAULT-14765?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: vault/activity/generation/generate_data.proto [hunk 1]
```diff
@@ -48,10 +48,9 @@ message Clients {
...
-  bool repeated = 4;
-  int32 repeated_from_month = 5;
-  string namespace = 6;
-  string mount = 7;
-  bool non_entity = 8;
+  bool repeated = 3;
+  int32 repeated_from_month = 4;
+  string namespace = 5;
+  string mount = 6;
+  bool non_entity = 7;
```
Comment (banks) [inline vault/activity/generation/generate_data.proto:55]: Renumbering message fields in protobuf seems like a thing we should probably never do (unless this was never used before etc. but that seems unlikely in a backport) that could cause back compatibility issues etc.

Was there a reason for this?

Comment (mpalmi) [reply vault/activity/generation/generate_data.proto:55]: This protobuf is new and only used to generate mock client counts for testing, so should be perfectly safe.

**Comment by mpalmi** [comment]
While going through the backport, there were 4 classes of commits (marked below). The backports for 1.11.x also pull in HLL estimate changes, which will be marked with ⏫ to show where this PR begins:

❌ empty commits, which had already been backported but weren't attributed by git
❌ superfluous changes which rejected the entire commit (these got the 'ol --skip during the cherry-pick)
[no modifier] changes that ported over (sometimes with amendment)
[no modifier] clean cherry-picks, which ported over with no modifications
⏫ HLL estimate backports

⏫ 2a8a8a8946 fix off by one error in activity log nil padding for month data (#15731)
⏫ 802ba7803e iterate through all available logs for precomputation and query gets (#15768)
⏫ 1575454ae6 refactor some code in modifyResponseMonths and ensure that the last mo… (#15767)
⏫ c88c73e508 Activity Log Filtering Limit Parameter (#16000)
⏫ cf60460d46 ActivityLog Implement HyperLogLog Store Functionality During Precomputation (#16146)
⏫ 2d1222630d activity log refactoring port (#16162)
⏫ 7e72219cfc Port: Use Stored Hll to Compute New Clients For Current Month (#16184)
⏫ 614cee3178 VAULT-6614 Enable role based quotas for lease-count quotas (OSS) (#16157)
⏫ 7140730885 Current Month Activity Estimate, Probabilistic Tests, And Bugfixes (#16447)
⏫ ff504adc52 re-add namespace attribution to current month (#16473)
⏫ 687b3d1c7f oss port of vault-7225-bugfix (#16745)
❌ 89f08de845 HCP link integration (#16939)
8dedb31a78 Make some activity log tests less flaky (#17028)
❌ 8258b36077 make proto (#17120)
❌ 72799f7bdf Update protos to match update of protobuf go library (#17215)
50a1f9e86a fix off by one err in current month client count computation (#17457)
❌ 0423ffbe24 update protoc version to 3.21.7 oss  (#17499)
❌ 483ef19946 Eslint prefer-const (#17864)
840abfbe10 Fix activity log end time (#17856)
87aa644ca8 Don't return a 204 if there's no historical data (#17935)
❌ 186ee31b10 Link OSS (#18228)
b8de2c2b4e De-duplicate namespaces when historical and current month data are mixed (#18452)
cb61488b9a Account for mount counts when de-duplicating current and historical month data (#18598)
d641bbc28e Add new clients into the monthly breakdown (#18629)
2a5a07e390 Revert "Add new clients into the monthly breakdown (#18629)" (#18726)
771bd8ba6d Add new clients into the monthly breakdown (#18766)
❌ 46b9921aae Allow Token Create Requests To Be Replicated (#18689)
❌ fbed961190 UI: combine current + history client count tabs into one dashboard (#17575)
b5d7d47ca2 VAULT-13061: Fix mount path discrepancy in activity log (#18916)
a9e17c2011 VAULT-13763 normalize activity log mount paths  (#19343)
❌ 7ef729708b UI: Use specific date in clients activity test (#19419)
9f7f8d5bfa VAULT-13729 activity log test godocs (#19433)
❌ e55c18ed12 adding copyright header (#19555)
e3c59773e9 Add no-op CensusAgent (#19625)
❌ e9d6dbce23 activitylog: Fix pq.Get trace logger output (#19650)
❌ bf2f1a88d6 UI: Fix flaky time-related tests (#19521)
b4fab6ac2a VAULT-13191: OSS changes (#19891)
❌ 54904e4cd6 VAULT-14733: Refactor processClientRecord in activity log (#19933)
d70c17f7fd VAULT-14733: SegmentReader interface for reading activity log segments (#19934)
4b6ec4079d Require activity log retention months at least the minimum (#20078)
05ba6bbddd api: Add reporting fields to activitylog config endpoint (#20086)
❌ c95d4fb15a VAULT-14734: activity log write endpoint (#20019)
❌ 75c903f0d6 openapi: Add display attributes for /sys (p2) (#19707)
002a59a370 Add minimum_retention_months to config endpoint (#20150)
77f83d9fe8 Refactor reporter for unseal setup (#20296)
730d0e2821 VAULT-14733: Split logic of precomputedQueryWorker (#20073)
❌ d234111801 Start counting ACME certificate issuance as client activity (#20520)
35e2c1665f VAULT-15703: Reload automated reporting (#20680)
❌ 6d95f8ca07 Add client_type field to EntityRecord protobuf (#20626)
810d504e4f Add current_billing_period activity endpoint param (#20694)
5b23dd506f VAULT-14735: generate mock clients for activity log  (#20252)
018ea84997 VAULT-15395: Support mocking time functions in the activity log (#20720)
541f18eeb7 VAULT-14735: repeated and segmented activity log clients (#20699)
b4e2751a09 VAULT-14735: write mock activity log entity files (#20702)
dc5dd71c72 Deflake TestActivityLog_MultipleFragmentsAndSegments (#20930)
5002489d27 VAULT-15394: Generate activity log precomputed queries (#20778)

File: vault/core.go [hunk 2]
```diff
@@ -3331,3 +3349,22 @@ func (c *Core) CheckPluginPerms(pluginName string) (err error) {
...
+		clone, err := entry.Clone()
+		if err != nil {
+			return nil, err
+		}
+
+		entries = append(entries, clone)
+	}
+
+	return entries, nil
+}
```
Comment (raskchanky) [inline vault/core.go:3370]: This doesn't look related to activity log stuff. Is this correct?

File: vault/logical_system_test.go [hunk 3]
```diff
@@ -4791,7 +4792,10 @@ func TestSystemBackend_LoggersByName(t *testing.T) {
 		t.Run(fmt.Sprintf("loggers-by-name-%s", tc.logger), func(t *testing.T) {
 			t.Parallel()
 
-			core, b, _ := testCoreSystemBackend(t)
+			core, _, _ := TestCoreUnsealedWithConfig(t, &CoreConfig{
+				Logger: logging.NewVaultLogger(hclog.Trace),
+			})
+			b := core.systemBackend
```
Comment (raskchanky) [inline vault/logical_system_test.go:4798]: This doesn't look related to activity log stuff. Is this correct?

File: vault/testing.go [hunk 4]
```diff
@@ -357,16 +361,21 @@ func TestCoreUnsealed(t testing.T) (*Core, [][]byte, string) {
...
-	inmemSink := metrics.NewInmemSink(1000000*time.Hour, 2000000*time.Hour)
 	conf := &CoreConfig{
 		BuiltinRegistry: NewMockBuiltinRegistry(),
-		MetricSink:      metricsutil.NewClusterMetricSink("test-cluster", inmemSink),
-		MetricsHelper:   metricsutil.NewMetricsHelper(inmemSink, false),
 	}
+	sink := SetupMetrics(conf)
 	core, keys, root := testCoreUnsealed(t, TestCoreWithSealAndUI(t, conf))
-	return core, keys, root, inmemSink
+	return core, keys, root, sink
```
Comment (raskchanky) [inline vault/testing.go:378]: Are all these metrics changes related to activity log/census things?

File: vault/core.go (hunk 2 above)
Comment (mpalmi) [reply vault/core.go:3370]: I had to pull this in (see: https://github.com/hashicorp/vault/pull/21210/commits/4dde07d5c6be66e58b66d9c307230b00066ed2c3) for one of the activitylog tests.

File: vault/logical_system_test.go (hunk 3 above)
Comment (mpalmi) [reply vault/logical_system_test.go:4798]: This is a part of the "Deflake" commit: 28dd651c771ee495902762589c41456659767302

File: command/server/config_test_helpers.go [hunk 5]
```diff
@@ -955,6 +955,7 @@ func testParseSeals(t *testing.T) {
 			},
 		},
 	}
+	addExpectedDefaultEntConfig(expected)
```
[the remark this answers is not in the document]
Comment (mpalmi) [reply command/server/config_test_helpers.go:958]: This is needed for config changes on the enterprise side. See https://github.com/hashicorp/vault/pull/21210/commits/05015942104ed5cc9fa3991acd7441f1a16abb4c

File: command/server/config_test_helpers_util.go [hunk 6]
```diff
@@ -3,4 +3,5 @@
 package server
 
 func addExpectedEntConfig(c *Config, sentinelModules []string)                         {}
+func addExpectedDefaultEntConfig(c *Config)                                            {}
```
[the remark this answers is not in the document]
Comment (mpalmi) [reply command/server/config_test_helpers_util.go:6]: Also related to https://github.com/hashicorp/vault/pull/21210/commits/05015942104ed5cc9fa3991acd7441f1a16abb4c

File: vault/testing.go (hunk 4 above)
Comment (mpalmi) [reply vault/testing.go:378]: This came along with 2eb6aa86b704b679b5db365f6cb4553d530a9b8b (backported from https://github.com/hashicorp/vault/commit/8dedb31a78223bd1790cdd09614371ba6ef08e8c)

**Review by raskchanky — APPROVED** [review]
